src/Catalog/Helpers/GalleryDatabaseQueryService.cs

```diff
@@ -0,0 +1,150 @@
+                }
+
+                packagesWithSameKeyDate.Add(package);
+            }
+
+            return result;
+        }
+
+        /// <summary>
+        /// Builds the SQL query string for db2catalog.
+        /// </summary>
+        /// <param name="cursor">The <see cref="DB2CatalogCursor"/> to be used.</param>
+        /// <returns>The SQL query string for the db2catalog job, build from the the provided <see cref="DB2CatalogCursor"/>.</returns>
+        internal static string BuildDB2CatalogSqlQuery(DB2CatalogCursor cursor)
+        {
+            return $@"SELECT TOP {cursor.Top} WITH TIES
+                        PR.[Id],
+                        P.[NormalizedVersion],
+                        P.[Created],
+                        P.[LastEdited],
+                        P.[Published],
+                        P.[Listed],
+                        P.[HideLicenseReport],
+                        P.[LicenseNames],
+                        P.[LicenseReportUrl]
+                    FROM [dbo].[Packages] AS P
+                    INNER JOIN [dbo].[PackageRegistrations] AS PR ON P.[PackageRegistrationKey] = PR.[Key]
+                    WHERE P.[PackageStatusKey] = {PackageStatus.Available} 
+                        AND P.[{cursor.ColumnName}] > @{CursorParameterName}
+                    ORDER BY P.[{cursor.ColumnName}]";
+        }
+
+        /// <summary>
+        /// Asynchronously gets a <see cref="IReadOnlyCollection{FeedPackageDetails}"/> from the gallery database.
+        /// </summary>
+        /// <param name="cursor">Defines the cursor to be used.</param>
+        /// <returns>A task that represents the asynchronous operation.
+        /// The task result (<see cref="Task{TResult}.Result" />) returns an
+        /// <see cref="IReadOnlyCollection{FeedPackageDetails}" />.</returns>
+        private async Task<IReadOnlyCollection<FeedPackageDetails>> GetPackages(DB2CatalogCursor cursor)
+        {
+            using (var sqlConnection = await _connectionFactory.OpenAsync())
+            {
+                return await GetPackageDetailsAsync(sqlConnection, cursor);
+            }
+        }
+
+        private async Task<IReadOnlyCollection<FeedPackageDetails>> GetPackageDetailsAsync(
+            SqlConnection sqlConnection,
+            DB2CatalogCursor cursor)
+        {
+            var packages = new List<FeedPackageDetails>();
+            var packageQuery = BuildDB2CatalogSqlQuery(cursor);
+
+            var packagesCommand = new SqlCommand(packageQuery, sqlConnection)
+            {
+                CommandTimeout = _commandTimeout
+            };
+
+            packagesCommand.Parameters.AddWithValue(CursorParameterName, cursor.CursorValue);
+
+            using (var packagesReader = await packagesCommand.ExecuteReaderAsync())
+            {
+                while (await packagesReader.ReadAsync())
+                {
+                    packages.Add(_db2catalogProjection.FromDataRecord(packagesReader));
```

Comment:
One note about this implementation--it ALWAYS grabs the next 20 packages with ties, even if the last timestamp has 10,000 packages sharing it. In the previous implementation, it would have grabbed the next 19 packages and then saved the last timestamp for a batch on its own. I'm not sure about the implications of this--if the 19 packages on their own are massive (e.g. have a lot of versions) this could wreck our pipeline. Could we make it so that it splits large timestamps into their own batch?

Reply:
Yes, that's indeed something to fix still. Are you suggesting to reintroduce the Group By in this single query?

Reply:
No, I'm suggesting that after we get the data from the server we filter it out--I think keeping the single query behavior is most important.

You could probably do it in [`OrderPackagesByKeyDate`](https://github.com/NuGet/NuGet.Services.Metadata/pull/524/files#diff-9238f95b1acac0dc5e93d264bfd09e36R65) somewhere.

Reply:
Addressed with https://github.com/NuGet/NuGet.Services.Metadata/pull/524/commits/dc4291d70161ac97bda66d952962c468eafbfede